### ISSUE TITLE
Add annotation-no-unknown

### ISSUE THREAD

**Issue by ssivanatarajan** [stylelint/stylelint#4508]
> Clearly describe the bug

"When there is a spelling mistake for an important declaration it is not reported as parse error"

> Which rule, if any, is the bug related to?

Parsing error

> What CSS is needed to reproduce the bug?

```css
.a{
color: red !imprtant;
}
```

> What stylelint configuration is needed to reproduce the bug?

`Since it is related to parsing, no need for the specific configuration.`

> Which version of stylelint are you using?

`12.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

 `Node API `

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

 `No`

> What did you expect to happen?

 `Parse error needs to be reported.`

> What actually happened (e.g. what warnings or errors did you get)?

 
`No errors reported`

**Comment by evilebottnawi** [comment]
It is out of scope stylelint, we don't check your syntax

**Comment by jeddy3** [comment]
@ssivanatarajan Thanks for the suggestion.

I'm going to reopen this issue for discussion.

As per [the VISION](https://stylelint.io/vision#possible-errors):

> Invalid code is best handled by emerging dedicated tools e.g. csstree - a language parser with syntax validation. __As a stop-gap, while these tools mature, provide invalid code rules for the simplest of cases.__

[csstree-validator](https://csstree.github.io/docs/validator.html) doesn't detect `!imprtant;`.

I think there's a valid argument for an `annotation-no-unknown` rule. ([Spec ref](https://www.w3.org/TR/css-cascade-3/#importance) for name). 

The likes of [stylelint-scss](https://github.com/kristerkari/stylelint-scss) can extend the rule for annotations like [`!default`](https://sass-lang.com/documentation/variables#default-values). They can make use of [`utils.checkAgainstRule`](https://stylelint.io/developer-guide/plugins#stylelintutilscheckagainstrule) as they already do for [`scss/at-rule-no-unknown`](https://github.com/kristerkari/stylelint-scss/tree/master/src/rules/at-rule-no-unknown).

Unless there are objections, I'll label this issue up accordingly?

**Comment by jeddy3** [comment]
I'm going to label this up as there are no objections.

- Name: `annotation-no-unknown`
- Primary option: `true`
- Secondary options: `ignoreAnnotations: []`
- Message: `Unexpected unknown annotation "${annotation}"`
- Description: "Disallow unknown annotations"
- Section: "Possible errors" -> "Annotation"

@ssivanatarajan Feel free to [contribute](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#creating-a-new-rule) this rule. You can use [`property-no-unknown`](https://github.com/stylelint/stylelint/tree/master/lib/rules/property-no-unknown) as a blue print.

**Comment by mattxwang** [comment]
I'd like to take this issue up if that's possible.

A clarifying question: on finding unknown annotations, looking at the spec ref you've linked:

> A declaration is important if it has a !important annotation as defined by [[css-syntax-3]](https://www.w3.org/TR/css-cascade-3/#biblio-css-syntax-3), i.e. if the last two (non-whitespace, non-comment) tokens in its value are the delimiter token ! followed by the identifier token important. All other declarations are normal (non-[important](https://www.w3.org/TR/css-cascade-3/#important)).

My thought process is to:

* walk each declaration
* check if the last two tokens are of the form `!____`
    * should I be creating a new utility to find annotations in a declaration? similar to [`getImportantPosition`](https://github.com/stylelint/stylelint/blob/main/lib/utils/getImportantPosition.js#L10)?
* if it is an unknown annotation (i.e. not `!important` or part of `ignoreAnnotations`), report an error

Is this line of thinking correct?

**Comment by ybiquitous** [comment]
@mattxwang Great for you to take it on!

For implementation, I think `postcss-value-parser` does an excellent job parsing tokens in a declaration's value.

E.g.

```console
$ node -p 'require("postcss-value-parser")("red !important")'
ValueParser {
  nodes: [
    { type: 'word', sourceIndex: 0, sourceEndIndex: 3, value: 'red' },
    { type: 'space', sourceIndex: 3, sourceEndIndex: 4, value: ' ' },
    {
      type: 'word',
      sourceIndex: 4,
      sourceEndIndex: 14,
      value: '!important'
    }
  ]
}
```

**Comment by jeddy3** [comment]
Yes, let's use the value parser to find annotations:

```js
function isAnnotation(node) {
  return node.type === 'word' && node.value.startsWith("!");
}
```

Let's return early if there's no `!` in the value string to avoid calling the value parser unnecessarily:

```js
if (!decl.value.includes('!')) return;
```

Let's add a new set `keywordSets.annotations` (of one) to [our references](https://github.com/stylelint/stylelint/blob/main/lib/reference/keywordSets.js):

```js
keywordSets.annotations = new Set(['!important']);
```

So:

1. walk each declaration
2. return early if no `!` in the value string 
3. use the value parser to find annotations (and their positions)
4. report if unknown and not ignored
